### ISSUE TITLE
madeira rayon color palette is weirdly broken/wrong

### ISSUE THREAD

**Issue by einsweniger** [inkstitch/inkstitch#3257]
The manufacturer says, their "Rayon 40" (semi-professional) line to consist of 180 colors, see: https://garne.madeira.de/1/downloads/colors/ (zip-files of all 180 thread colors, photographed, easily sortable by color code)
The `InkStitch Madeira Rayon.gpl` contains over 350 colors, and except for black (1000) and white (1001), none of the color names match my box of over 80 colors. 

On the other hand, there's a list of 422 "Classic 40" (professional line) colors, which seem to be a super-set of the rayon 40 line, so for all intents and purposes, I'd say most of the Color Palettes for Madeira Rayon are broken, which could quite possibly be related to https://github.com/inkstitch/inkstitch/issues/2668 

There's reference cards for both lines, bot they only list the color code, without a name. I only found a mapping for 120 color code -> name, that all match the names in my thread box. There's also an official pantone -> classic 40 mapping (unfortunately, not bijective).

I'm not sure where the data was originally sourced from or at what point they turned sour, but I'd rather have the color palettes for madeira remade. 

If the names are important, I could ask the manufacturer if they have a list of code <-> name, or if such a thing even exists. If not, we could check if the Pantone reference codes have named colors.

I'll remake them, because this annoys me greatly, I'll provide them to this project as well, If you're Interested. I just wanted to have the bug report here, so that others might see the small little caveat for color conversion.

References:
 - Classic 40: https://www.madeira.com/fileadmin/user_upload/Downloads/Shade_Cards/MADEIRA_CLASSIC.pdf
 - Rayon 40: https://garne.madeira.de/fileadmin/user_upload/home/farbkarten/Colour_Card_RAYON.pdf
 - Pantone -> Classic 40: https://shop.madeira.com/data/media/docs/CLASSIC-PANTONE-CONVERSION-CHART-09-2014_01.pdf

**Comment by einsweniger** [comment]
A little progress report. 

I just talked to the german manufacturers about providing names for the colors.
some background: Madeira is actually two different companies: “MADEIRA Garnfabrik” and “MADEIRA Garne”.
_Garne_ is in charge of the “Rayon” line, and _Garnfabrik_ does the “Classic” line.

_Garnfabrik_ told me pretty straight forward, that they only have codes and no names for their +400 color Classic lineup.

_Garne_ on the other hand told me that they only made those names for the NA market, and that they officially have no names as well. They did however, help me navigate their website, where I could finally find a chart for at least 140 of the 180 colors, located [here](https://garne.madeira.de/fileadmin/user_upload/home/farbstreifen/Farbtabellen_2023/Rayon_No40_200m_Farbtabelle.jpg)

so, basically, for rayon, I have to make up about 40 names for colors and script that to create a new palette, or maybe I’m in luck and can find the missing one buried somewhere on the madeira website.

**Comment by einsweniger** [comment]
(Another thought, that I had: I’ve recently got a firmware update file for my stitching machine, this probably contains some names for the colors as well)

**Comment by tyrosinase** [comment]
Possibly further confusing the issue: Urban Threads/Embroidery Library uses Madeira rayon (IIRC) for their official thread palettes, and the names never match what I find when I've bought Madeira.

They've got a thread conversion thingy but doesn't look like it has names in it at all, so I'm not sure they've got a general listing of it but I could probably parse them out of the thread listings pretty easily, along with the hex codes. https://urbanthreads.com/thread-exchange

(I use mostly Sulky rayon, so the multiple names are just a little extra complication on top of conversion issues)

**Comment by kaalleen** [comment]
I had the same problem for the thread that I use and made my palette with color codes only. So I simply use the code twice in the palette - once for the code itself and then also for the name. Maybe I do not see the point, but I'm not sure what would be the use of making names up which won't help to identify the actual thread. On the other hand it might be nice to see that this color is a kind of blue or something...

**Comment by LochanLoch** [comment]
And how do you want to depict the multi-colored yarn with one color... unfortunately that's not possible. 
Had the same problem but it is not realistic to display (even expensive software has its limits too) or you can look at Coloreel.

**Comment by kaalleen** [comment]
@einsweniger On second though, making up names could help blind people (we had requests in the past) to get an idea of how the color would look like, so this would oppose the idea to just use the color code as a name as well. What do you think would be a good approach?

**Comment by tyrosinase** [comment]
*raises hand* Color names are definitely useful; one of my apprentices (read: my kid) is partially colorblind, and knowing "this looks brown but the name says it's blue" or whatever helps sometimes.

**Comment by einsweniger** [comment]
Hey y’all, thank you for the input. I had this on the back-burner for a while, but I’m still willing to do some work here. I’d really like to provide some names, especially in the interest for accessibility. I’ll check if I can use the pantone color mapping for names. if not, I’ll check if there is an agreed upon (quasi-)standard (whatwg, maybe? Or RAL, would be another possibility) for color names, that I could use instead. Haven’t checked, yet if such a thing exists  
For the yarns with multiple thread colors, I’d add a remark of some sort to that color name. I’d be interested in your feedback, before I start working on it.

**Comment by einsweniger** [comment]
Pantone is out of the question, they don’t like sharing 🤷

**Comment by kaalleen** [comment]
I'd be great if you could add the color names then. I'm afraid though, we do not have a standard yet. A lot of the color palettes were provided by X3msnake a long time ago. Some were corrected by other contributors later on.
Some of the rgb color definitions - I don't remember which ones - were created with a physical color sensor.

So I suppose everything is a bit mixed up.

**Comment by einsweniger** [comment]
So, I went down a rabbit hole the last few days, many things are broken. I don’t think this is a inkstitch issue, but I discovered another few in the process that are. First the (I think non-issues):
My brother machine just ignores the colors that are set, even when forcing the export of PESv6 (through some hack), my machine just ignores the colors/codes/manufacturer id that are set, I guess brother have their own logic on how to interpret that info. Even when feeding that machine a T(something) file (that compatible format), it’s ignored. Even when exporting enhanced headers (a hack, again). Not sure if that’s a problem in pyemb or if it’s machine specific. 


Aaaanyway, I built a custom palette, but I could not apply it. I had to edit the inx file that presents the options, and put my palette in. That’s one of the ink/stitch things. The other are mostly related to the options that you could apply when exporting to a sewing format: pyemb supports rendering the T format with extended headers, as well as PES with something else that v1, currently only v6, because pf all the unknowns. 
just a heads up, that this continues, I’ll put in sum PR/issues when I have a little more time.

**Comment by kaalleen** [comment]
Thanks for the heads up.

The pec color set can be found here and is possibly your problem (pes is a bit specific with the colors I guess): https://github.com/inkstitch/pyembroidery/blob/main/pyembroidery/EmbThreadPec.py
I never really worked a lot on pyembroidery. So my knowledge is a bit limited in this area.

> Aaaanyway, I built a custom palette, but I could not apply it. I had to edit the inx file that presents the options, and put my palette in. That’s one of the ink/stitch things. 

I think I solved this issue for the next release. It's a wxpython extension now. So it can also read user palettes in when it is called.

**Comment by einsweniger** [comment]
> I think I solved this issue for the next release. It's a wxpython extension now. So it can also read user palettes in when it is called.

That's very cool! Thank you! 

Regarding the (64) pec colors: yes, that's what I gathered as well. I also got down into the nitty gritty with some files published by brother, which pyemb does not parse quite right (no gripes here, that format is sparsely documented) which with v8/v10 seems (!?) to be able to encode arbitrary RGB values and displayed color codes, but that might also be down to specific hardware models and (maybe) respective firmware versions of that hardware, on how they interpret all of that information. Or what they are limited/able to display on that specific hardware. Most of this comes down to how pyemb renders that information or how the brother machines are programmed to interpret that information. I'll try to contact the manufacturer over the next few days, maybe they want to share some information about their formats. Which I could then port to pyemb/inkstitch.
